### ISSUE TITLE
Support starlette "lifespan" context for application

### ISSUE THREAD

**Issue by uSpike** [tiangolo/fastapi#2943]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

Starlette added support for the lifespan context function in 0.13.5.  It's not well documented, but the code does [suggest](https://github.com/encode/starlette/blob/ed73b97c9b8f4aa527eed88032c04ce9fe109a3c/starlette/applications.py#L51):
```python
        # The lifespan context function is a newer style that replaces
        # on_startup / on_shutdown handlers. Use one or the other, not both.
```

For my purposes, it's much more convenient to use a (async)contextmanager for the startup and shutdown events of my application.  It would be nice to have the option

### The solution you would like

I'd like an easier way, and one that's officially supported, to use the lifespan context function.

```Python
from fastapi import FastAPI

async def lifespan(app):
    print("startup")
    async with SomeResource():
        yield
    print("shutdown")

app = FastAPI(lifespan=lifespan)
```

or

```Python
from fastapi import FastAPI

app = FastAPI()

@app.lifespan
async def lifespan(app):
    print("startup")
    async with SomeResource():
        yield
    print("shutdown")
```

### Describe alternatives you've considered

I can already accomplish this simply by doing:
```python
from fastapi import FastAPI

async def lifespan(app):
    print("startup")
    async with SomeResource():
        yield
    print("shutdown")

app = FastAPI()
app.router.lifespan_context = lifespan
```
however this is not officially supported and would likely break if accidentally using `app.on_event` in addition.

One could also do nasty stuff with `__aenter__` and `__aexit__`:
```python
from fastapi import FastAPI

app = FastAPI()

@app.on_event("startup")
async def startup()
    print("startup")
    app.state.resource = SomeResource()
    await app.state.resource.__aenter__()

@app.on_event("shutdown")
async def startup()
    print("shutdown")
    await app.state.resource.__aexit__(None, None, None)
```
but that seems quite ugly to me.

### Environment

* OS: Linux
* FastAPI Version: 0.63.0
* Python version: 3.8.5

**Comment by juntatalor** [comment]
@uSpike Hi! Did you look into [asgi-lifespan](https://pypi.org/project/asgi-lifespan/) package?

**Comment by uSpike** [comment]
@juntatalor I did not look in to asgi-lifespan.  It seems strange to use something outside of starlette, when starlette alreawdy supports the lifespan context:

```python
from starlette import Starlette

async def lifespan(app):
    async with SomeResource():
        yield

app = Starlette(lifespan=lifespan)
```

What I'm looking for in this request is to have the starlette functionality exposed in fastapi.

**Comment by sm-Fifteen** [comment]
See also issue #617 (which is about potential ways to use lifespan with FastAPI's dependency injection system and potential APIs for that) and encode/starlette#799 (which implemented the lifespan context manager).

**Comment by tiangolo** [comment]
Thank you @uSpike! And thanks for the PR implementing it. :rocket: :cake: 

This will be available in FastAPI `0.93.0`, released in the next few hours. :tada: 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by Soberia** [comment]
Not having the `app = FastAPI()` line at the top of the module is such a pain!
`lifespan` also needs to be provided as a decorator.

```python
@app.lifespan
async def lifespan(app: FastAPI) -> None:
    ...
```
